llvm/lib/Transforms/IPO/ConstantMerge.cpp

```diff
@@ -226,9 +226,7 @@ static bool mergeConstants(Module &M) {
     // Now that we have figured out which replacements must be made, do them all
     // now.  This avoid invalidating the pointers in CMap, which are unneeded
     // now.
-    for (unsigned i = 0, e = SameContentReplacements.size(); i != e; ++i) {
-      GlobalVariable *Old = SameContentReplacements[i].first;
-      GlobalVariable *New = SameContentReplacements[i].second;
+    for (const auto &[Old, New] : SameContentReplacements) {
```

Comment:
No &

Reply:
I just tried to drop `&`, but that resulted in:

```
llvm/lib/Transforms/IPO/ConstantMerge.cpp:229:21: error: loop variable '[Old, New]'
creates a copy from type 'std::pair<llvm::GlobalVariable *, llvm::GlobalVariable *>
const' [-Werror,-Wrange-loop-construct]
```

Reply:
I don't understand this warning. It sounds useless

Reply:
Probably designed to help catch cases of copying large structures. Two words seems a bit small to trigger such a warning. I think I've pushed back on this elsewhere too.

It doesn't look like this is a clang warning? ( https://godbolt.org/z/sKfKsYj8b ) - perhaps it's some clang-tidy diagnosed issue?

Where'd you see this warning, @kazutakahirata ?

Reply:
@dwblaikie I see the warning with clang-16.0.6.

Here is a reduced testcase: https://godbolt.org/z/4Eo43hb81

I tried clang-18.1.0, clang-17.0.1, and clang-16.0.0 on Compiler Explorer.  clang-16 and clang-17 issue the warning while clang-18 doesn't.  Probably, the compiler was fixed sometime between clang-17 and clang-18.

Reply:
Ah, looks like https://github.com/llvm/llvm-project/issues/47355 - std::pair is trivially copy constructible, but isn't trivially copy assignable (& so is not "trivially copyable") - and the warning used "trivially copyable" instead of "trivially copy constructible" for the test, so warned in this case - because it thought this was invoking a non-trivial copy constructor. So regardless of how small the pair is, it'd warn.

Do we have this warning on by default in llvm builds? Or is it turned on some other way in your dev environment?

It might be worth turning off this warning (maybe, if someone can be bothered, turning it off only if it's buggy in this way - we do have other warnings we have little compatibility tests for in cmake that check the warning doesn't have certain bugs before enabling it) due to this bug.